server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/HierarchicalKMeans.java

```diff
@@ -0,0 +1,194 @@
+            float f = Math.min((float) samplesPerCluster / targetSize, 1.0f);
+            int localSampleSize = (int) (f * vectors.size());
+            KMeansLocal kMeansLocal = new KMeansLocal(localSampleSize, maxIterations, clustersPerNeighborhood);
+            kMeansLocal.cluster(vectors, kMeansResult);
+        }
+
+        return kMeansResult;
+    }
+
+    KMeansResult kMeansHierarchical(final FloatVectorValuesSlice vectors, final int targetSize) throws IOException {
+        if (vectors.size() <= targetSize) {
+            return new KMeansResult();
+        }
+
+        int k = Math.clamp((int) ((vectors.size() + targetSize / 2.0f) / (float) targetSize), 2, MAXK);
+        int m = Math.min(k * samplesPerCluster, vectors.size());
+
+        // TODO: instead of creating a sub-cluster assignments reuse the parent array each time
+        int[] assignments = new int[vectors.size()];
+
+        KMeans kmeans = new KMeans(m, maxIterations);
+        float[][] centroids = KMeans.pickInitialCentroids(vectors, k);
+        KMeansResult kMeansResult = new KMeansResult(centroids);
+        kmeans.cluster(vectors, kMeansResult);
+
+        int[] clusterSizes = new int[centroids.length];
+
+        // TODO: consider adding cluster size counts to the kmeans algo
+        // handle assignment here so we can track distance and cluster size
+        int[] centroidVectorCount = new int[centroids.length];
+        float[][] nextCentroids = new float[centroids.length][vectors.dimension()];
+        for (int i = 0; i < vectors.size(); i++) {
+            float smallest = Float.MAX_VALUE;
+            int centroidIdx = -1;
+            float[] vector = vectors.vectorValue(i);
+            for (int j = 0; j < centroids.length; j++) {
+                float[] centroid = centroids[j];
+                float d = VectorUtil.squareDistance(vector, centroid);
+                if (d < smallest) {
+                    smallest = d;
+                    centroidIdx = j;
+                }
+            }
+            centroidVectorCount[centroidIdx]++;
+            for (int j = 0; j < vectors.dimension(); j++) {
+                nextCentroids[centroidIdx][j] += vector[j];
+            }
+            assignments[i] = centroidIdx;
+            clusterSizes[centroidIdx]++;
+        }
+
+        // update centroids based on assignments of all vectors
+        for (int i = 0; i < centroids.length; i++) {
+            if (centroidVectorCount[i] > 0) {
+                for (int j = 0; j < vectors.dimension(); j++) {
+                    centroids[i][j] = nextCentroids[i][j] / centroidVectorCount[i];
+                }
+            }
+        }
+
+        int effectiveK = 0;
+        for (int i = 0; i < clusterSizes.length; i++) {
+            if (clusterSizes[i] > 0) {
+                effectiveK++;
+            }
+        }
+
+        int[] assignmentOrdinals = vectors.slice();
```

Comment:
Instead of "assignment ordinals" we can have a `IntToInt` function that translates the ordinals appropriately and use the `ordToDocId` interface already provided in the FloatVectorValue class.